examples/autograd/train.py

```diff
@@ -0,0 +1,257 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from singa import singa_wrap as singa
+from singa import opt
+from singa import device
+from singa import tensor
+import numpy as np
+import time
+import argparse
+from PIL import Image
+
+
+# Data Augmentation
+def augmentation(x, batch_size):
+    xpad = np.pad(x, [[0, 0], [0, 0], [4, 4], [4, 4]], 'symmetric')
+    for data_num in range(0, batch_size):
+        offset = np.random.randint(8, size=2)
+        x[data_num, :, :, :] = xpad[data_num, :,
+                                    offset[0]:offset[0] + x.shape[2],
+                                    offset[1]:offset[1] + x.shape[2]]
+        if_flip = np.random.randint(2)
+        if (if_flip):
+            x[data_num, :, :, :] = x[data_num, :, :, ::-1]
+    return x
+
+
+# Calculate Accuracy
+def accuracy(pred, target):
+    # y is binary array for ground truth
+    y = np.argmax(pred, axis=1)
+    # t is integer for ground truth, the dim of t is (batch, 1)
+    t = np.max(target, axis=1)
+    a = y == t
+    correct = np.array(a, "int").sum()
+    # print(correct)
+    return correct
+
+
+# Data partition according to the rank
+def partition(worker_id, num_workers, train_x, train_y, val_x, val_y):
+    # Partition training data
+    data_per_rank = train_x.shape[0] // num_workers
+    idx_start = worker_id * data_per_rank
+    idx_end = (worker_id + 1) * data_per_rank
+    train_x = train_x[idx_start:idx_end]
+    train_y = train_y[idx_start:idx_end]
+    # Partition evaluation data
+    data_per_rank = val_x.shape[0] // num_workers
+    idx_start = worker_id * data_per_rank
+    idx_end = (worker_id + 1) * data_per_rank
+    val_x = val_x[idx_start:idx_end]
+    val_y = val_y[idx_start:idx_end]
+    return train_x, train_y, val_x, val_y
+
+
+# Function to all reduce NUMPY Accuracy and Loss from Multiple Devices
+def reduce_variable(variable, dist_opt, reducer):
+    reducer.copy_from_numpy(variable)
+    dist_opt.all_reduce(reducer.data)
+    dist_opt.wait()
+    output = tensor.to_numpy(reducer)
+    return output
+
+
+def resize_dataset(x, image_size):
+    num_data = x.shape[0]
+    dim = x.shape[1]
+    X = np.zeros(shape=(num_data, dim, image_size, image_size), dtype=np.float32)
+    for n in range(0, num_data):
+        for d in range(0, dim):
+            X[n, d, :, :] = np.array(Image.fromarray(x[n, d, :, :]).resize(
+                (image_size, image_size), Image.BILINEAR),
+                                     dtype=np.float32)
+    return X
+
+
+def run(worker_id, num_workers, device_id, max_epoch, batch_size, model, data,
+        sgd):
+    dev = device.create_cuda_gpu_on(device_id)
+    dev.SetRandSeed(0)
+    np.random.seed(0)
+
+    if model == 'resnet':
+        from model import resnet
+        model = resnet.resnet50()
```

Comment:
can we run resnet50 over mnist and cifar10? the input size is very small.
after many pooling/conv layers, the feature map size may be equal or smaller than 1.

Reply:
can run resnet 50 on cifar10, I rescaled to 224 (see the resize function after data augmentation)

Reply:
does it affect the accuracy?
we can use resnet18 I think, which is also available.

Reply:
eval accuracy can achieve around 90% on resnet50, I will try resent 18